util/topsql/topsql.go

```diff
@@ -79,25 +79,33 @@ func Close() {
 	stmtstats.CloseAggregator()
 }
 
-// AttachSQLInfo attach the sql information info top sql.
-func AttachSQLInfo(ctx context.Context, normalizedSQL string, sqlDigest *parser.Digest, normalizedPlan string, planDigest *parser.Digest, isInternal bool) context.Context {
-	if len(normalizedSQL) == 0 || sqlDigest == nil || len(sqlDigest.Bytes()) == 0 {
-		return ctx
+// RegisterSQL uses to register SQL information into Top SQL.
+func RegisterSQL(normalizedSQL string, sqlDigest *parser.Digest, isInternal bool) {
+	if sqlDigest != nil {
+		sqlDigestBytes := sqlDigest.Bytes()
+		linkSQLTextWithDigest(sqlDigestBytes, normalizedSQL, isInternal)
 	}
-	var sqlDigestBytes, planDigestBytes []byte
-	sqlDigestBytes = sqlDigest.Bytes()
+}
+
+// RegisterPlan uses to register plan information into Top SQL.
+func RegisterPlan(normalizedPlan string, planDigest *parser.Digest) {
 	if planDigest != nil {
-		planDigestBytes = planDigest.Bytes()
+		planDigestBytes := planDigest.Bytes()
+		linkPlanTextWithDigest(planDigestBytes, normalizedPlan)
 	}
-	ctx = collector.CtxWithDigest(ctx, sqlDigestBytes, planDigestBytes)
-	pprof.SetGoroutineLabels(ctx)
+}
 
-	if len(normalizedPlan) == 0 || len(planDigestBytes) == 0 {
-		// If plan digest is '', indicate it is the first time to attach the SQL info, since it only know the sql digest.
-		linkSQLTextWithDigest(sqlDigestBytes, normalizedSQL, isInternal)
-	} else {
-		linkPlanTextWithDigest(planDigestBytes, normalizedPlan)
+// AttachSQLInfo attach the sql information into Top SQL and register the SQL meta information.
+func AttachSQLInfo(ctx context.Context, normalizedSQL string, sqlDigest *parser.Digest, isInternal bool) context.Context {
+	if sqlDigest == nil || len(sqlDigest.Bytes()) == 0 {
+		return ctx
 	}
+	sqlDigestBytes := sqlDigest.Bytes()
+	ctx = collector.CtxWithSQLDigest(ctx, sqlDigestBytes)
+	pprof.SetGoroutineLabels(ctx)
+
+	linkSQLTextWithDigest(sqlDigestBytes, normalizedSQL, isInternal)
```

Comment:
duplicated with RegisterSQL?

Reply:
No, if this function is called, then `RegisterSQL` won't be call anymore.